geps/gep-1651.md

```diff
@@ -0,0 +1,218 @@
+  are only accessible within a single cluster
+- Define a `Public` scope for addresses that are accessible to the internet
+- Implementations can define & support their own scopes
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "internal"
+(service.namespace.svc) and "external" (service.namespace.example.com).
+Gateways using the same GatewayClass on the cluster, but ensure that the
+"internal" service is only reachable within the cluster. This would greatly
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of
+layer 7 load balancing for our internal services.
+
+## API
+
+We propose adding a new `reachability` field under the `spec.infrastructure` stanza of a Gateway.
+
+### Predefined Reachability Values
+
+Implementations MAY implement the following values for 'reachability' and MUST abide by
+their defined semantics.
+
+Value | Semantic
+-|-
+`public`|The address is reachability over the public internet
+`cluster`|The address is only reachability within the bounds defined by the [cluster's network model](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+
+### Vendor prefixed values
+
+Implementations can define custom 'reachability' values by specifying a vendor prefix followed
+by a slash `/` and a custom name ie. `com.ingress.dev/my-reachability`.
+
+### Default reachability
+The default value of `reachability` is implementation specific. It is RECOMMENDED that
+the default `reachability` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+### Go
+
+```go
+
+type GatewayReachability string
+
+const (
+  GatewayReachabilityPublic  GatewayReachability = "public"
+  GatewayReachabilityCluster GatewayReachability = "cluster"
+)
+
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+type GatewayInfrastructure struct {
+  // AddressScope allows the Gateway to specify the accessibility of it's addresseses
+  Reachability GatewayReachability `json:"reachability"`
+}
+```
+
+### YAML
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  infrastructure:
+    reachability: public
+  listeners:
+  - protocol: HTTP
+    port: 80
+```
+
+### Semantics
+
+#### Unsupported reachability & address values
+
+If a Gateway is unable provide an address for the desired reachability it MUST set the condition `Accepted` to `False`
+
+### Status.Addresses
```

Comment:
Nit: Is this meant to be on the same level as Semantics? It feels like it might belong on the same level as "Unsupported reachability & address values" instead.

Reply:
Yup - I'll fix that.